### PR TITLE
feature(websocket): Add websocket client

### PR DESCRIPTION
**Pull request by Mohamedemad4**
## Summary of Changes
added a simplified Barebones interface to Tiingo's websocket API
implemented #2 

Just a couple of notes:
  - I didn't write tests cause i didn't know what to test for since this is a really simple enhancement 

and question.
what should I write in the ```HISTORY.rst``` exactly? I didn't understand that part.

@hydrosquall

### CODE REVIEW

File: tiingo/wsclient.py
Comment (hydrosquall) [inline tiingo/wsclient.py:30]: Hey, thanks for investigating this! I'm a little bit concerned by the use of `global` here, so I went and looked at the issue in the comment, and I'm not sure that I entirely follow why resolving this requires using globals. https://github.com/websocket-client/websocket-client/issues/471

I was wondering if there were any existing Websocket clients that this implementation was inspired by that I could take a look at.

File: tiingo/wsclient.py
Comment (hydrosquall) [inline tiingo/wsclient.py:3]: I was wondering, what is this try-except import for?

File: tiingo/wsclient.py
Comment (hydrosquall) [inline tiingo/wsclient.py:52]: What do you think about documenting the expected.permitted options for the config dictionary + the expected values for endpoint + the function signature for `on_msg_cb` ? The example code you provided in the docstring is a good start for a documentation page, but we would want to aim for something closer to a descriptive docstring for a class comment. See https://realpython.com/python-pep8/#documentation-strings for more background.

File: tiingo/wsclient.py
Comment (hydrosquall) [inline tiingo/wsclient.py:67]: I think `ticks` might have actually be `takes`

File: tiingo/wsclient.py
Comment (hydrosquall) [inline tiingo/wsclient.py:80]: I think we can use a `MissingRequiredArgument` error here, and perhaps describe what "on_msg_cb" means, as a new user may not know that cb is shorthand for "callback".

https://github.com/hydrosquall/tiingo-python/blob/master/tiingo/exceptions.py

**Review by hydrosquall — COMMENTED** [review]
Hi @Mohamedemad4 , thanks for taking the time to make this contribution! I have 3 questions, since it's been a while since we've added a new dependency to the library, so I want to make sure we do this carefully.

1.  What do you see are any pros/cons of choosing `websocket-client` over other choices?
2. Once someone has instantiated a TiingoWebsocketClient, will they ever do anything with the returned object, or does all of their logic have to be in the callback function?
3. For the purpose of documentation, could we try to describe the shape of the data returned in `msg` (or link to where someone can go to find out)?

**Comment by hydrosquall** [comment]
> I didn't write tests cause i didn't know what to test for since this is a really simple enhancement

I'll try this out on the weekend and let you know if anything comes to mind, but that's probably fine. We don't want to just be testing the underlying library, I suppose a test could be written for a function that validates the arguments passed into the constructor for the client if you'd like to practice getting that experience. It would look something like this:

https://github.com/hydrosquall/tiingo-python/blob/db92c33410f929c2cda27bdf1e7a905c008f3b8a/tests/test_tiingo.py#L29-L37

> what should I write in the HISTORY.rst exactly? I didn't understand that part.

This is a description of the changes that your PR is introducing - if you're not sure, you can also provide a suggestion directly as a PR comment, and I can update History.rst directly when I cut a new release.

**Comment by Mohamedemad4** [comment]
Thank you @hydrosquall  for taking the time to review my noob code.
as for your questions.

1- it appeared to be the most popular one. maintained and mature client
2- no they once the object is called all user logic should be written in the callback function
3- see commit febcd8a5f313a452f89c04317697617bcc0c8012

File: tiingo/wsclient.py
Comment (Mohamedemad4) [reply tiingo/wsclient.py:3]: to import the correct version of ```thread``` regardless of python version

File: tiingo/wsclient.py
Comment (Mohamedemad4) [reply tiingo/wsclient.py:31]: i have tracked this down to this [line of code](https://github.com/websocket-client/websocket-client/blob/29c15714ac9f5272e1adefc9c99b83420b409f63/websocket/_app.py#L343) in the  library.
which checks if the method is bounded before passing ```self``` to it which is needed for the number of arguments to match up. and ofc you can't have unbounded method access class data. I would love not to use ```global``` ,but i have no idea how to implement something like that without actually changing the ```websocket-client``` code. if you have an idea or if i am missing something, Please. let me know

File: tiingo/wsclient.py
Comment (hydrosquall) [inline tiingo/wsclient.py:88]: Instead of doing this try/except block here are raising a KeyError that gets turned into an Attribute error, what do you think about making a single if-statement which raises a single `InvalidArgumentError` (or just `Attribute` error if you don't want to create a new exception class)?

File: tiingo/wsclient.py
Comment (hydrosquall) [inline tiingo/wsclient.py:100]: Nice idea to put an example function in the error message for how to fix the bug, developers really appreciate that! :)

**Review by hydrosquall — CHANGES_REQUESTED** [review]
This is great progress, thanks for working with me on iterating on this. Could we try adding a fixture test to check for the API key + valid initialization arguments so that this client file file has some test coverage?

File: README.rst
Comment (Mohamedemad4) [reply README.rst:153]: I don't think adding examples would work out for stylistic reasons. anyways the link provided gives a lot of in-depth details

File: tiingo/wsclient.py
Comment (Mohamedemad4) [reply tiingo/wsclient.py:100]: glad you liked it

File: tiingo/wsclient.py
Comment (hydrosquall) [reply tiingo/wsclient.py:31]: I guess what I'm unsure about is why `genericWebsocketClient` class is necessary in the first place, since we never actually even read the value in the `ws_client` variable. This is why I was curious if this is based on an existing production library / program that you saw, so I can evaluate whether that approach makes sense for us too.

If we're not actually going to read the value of `ws_client`, I'm not sure that I see the value of creating it in the first place.

**Review by hydrosquall — COMMENTED** [review]
I was reviewing the results on https://lgtm.com/projects/g/hydrosquall/tiingo-python/rev/pr-598152663398970811215859fcbae6dc034640ef , and realized that I'm not sure about the role that `genericWebsocketClient` is serving. Would removing this class remove the need for creating global variables?

**Comment by Mohamedemad4** [comment]
```genericWebsocketClient``` is not necessary. BUT otherwise the functions inside it would have to be left in the global space of the file. I just thought it was neater to put it all in a class with an issue description.

**Comment by hydrosquall** [comment]
@Mohamedemad4 - thanks for the explanation. I see the intention behind using the class as a way to help keep the code organized. With that said, I see two advantages to avoiding the need creating a class:

- We no longer need to create a class with methods where `self` is not the first argument, which is unconventional in the python world (as you already knew since you left a comment explaining why it was necessary to do something unusual)
- We won't need to create globals: the callbacks and functions could be created just within scope of the `TiingoWebsocketClient`'s `__init__` method. 

Given that avoiding the class would prevent the need for globals and unconventional use of "self", I think this code might be easier to maintain if we omit the abstraction layer for now. What do you think?

**Comment by Mohamedemad4** [comment]
@hydrosquall 
If i understand what you are saying correctly. By globals you are referring to ```GLOB_config``` and ```GLOB_on_msg_cb```. right?
because in that case we would still need them simply because ```TiingoWebsocketClient``` *CANNOT* be passed to the callback functions via the ```self``` parameter.
However if you mean just putting the Callback methods in ```TiingoWebsocketClient``` scope,then i see no problems with that.

**Comment by hydrosquall** [comment]
> By globals you are referring to GLOB_config and GLOB_on_msg_cb. right?

Yes - I think if we put those callbacks directly in `TiingoWebsocketClient`'s init, it'll avoid the need for globals. I understand the original motivation to help keep the code organized by putting all the callbacks in a class, but in this case I'm not sure it helps us very much + risks confusing people with the conventional use of self + globals.

**Comment by Mohamedemad4** [comment]
@hydrosquall 
I am confused. do you mean the Callback methods should be put in TiingoWebsocketClient scope? 
because I see why we can do that.
BUT the whole point of the Globals is because you can't pass ```TiingoWebsocketClient``` instance to the callback when calling them.
so that's why ```GLOB_config``` and ```GLOB_on_msg_cb``` need to be global.

**Comment by hydrosquall** [comment]
> BUT the whole point of the Globals is because you can't pass `TiingoWebsocketClient` instance to the callback when calling them.

Hey, thanks for the additional explanation. Can you show me an example of where the callback would need access to the `TiingoWebsocketClient` instance? I don't think it's necessary if the callback only ever needs to know about the `message` item, and that can be done without creating a global.

I do see the one place where the global is used in the callback, however, it looks to me like that can be used without ever converting  the local `config` into a global.

https://github.com/hydrosquall/tiingo-python/blob/fa6b010e9a52975445b04799c1a292233df78b03/tiingo/wsclient.py#L28

**Comment by Mohamedemad4** [comment]
Well. I can see why you would want to get rid of the Globals.
However. the ```config``` dict needs to be passed to the ```on_open``` callback in order for Tiingo to know information about the client. and there is just no way of getting around to having to pass that ```config``` to the ```on_open``` callback (well maybe if you use something like ```time.sleep``` after opening the connection. but that's even worse.

And looking back at the code. You are correct. ```GLOB_on_msg_cb``` doesn't need to be global. strictly speaking at least.
But as for ```GLOB_config``` i really don't see it. 
plus having everything related to the callbacks in one place, should really help with readability and maintainability.
What do you think?

**Comment by hydrosquall** [comment]
> However. the config dict needs to be passed to the on_open callback in order for Tiingo to know information about the client. and there is just no way of getting around to having to pass that config to the on_open callback

How do you feel about using a higher order function? In case you haven't run into this before, it's a function that returns a function. This lets us "bake in" the config without needing to create a throwaway class (since we're never using that `ws_client` variable for anything. 

```python
def get_on_open(config):
          def on_open(ws):
               # ... everything the same as before but without globals
               def run(*args):
                   print(GLOB_config)
                   ws.send(json.dumps(GLOB_config))
              thread.start_new_thread(run, ())

         return on_open

def __init__(self,config={},endpoint=None,on_msg_cb=None):
     # ... same as before
    ws = websocket.WebSocketApp("ws://echo.websocket.org/",
                              on_message = on_message,
                              on_error = on_error,
                              on_close = get_on_open(config))
 
```

I really appreciate that you're thinking about maintainability in grouping these callbacks together under a class. However, I would say that I think the value of grouping all the callbacks together under 1 class is limited if the class isn't stateful/if the values of the class instance are never read themselves. I'd suggest waiting until we have a reason to actually read the value of the `ws_client` variable before creating a class to wrap all of those callbacks. 

Thanks for sharing more about your thought process for this and sticking with this PR! Let me know if you have any other questions, it has been a good learning experience for me going through this architecture/design decision with you.

**Comment by Mohamedemad4** [comment]
Thank you @hydrosquall for taking the time with me and rejecting it outright. Using Higher Order functions seems like a better solution.
will Commit the new code so you can review it

**Comment by Mohamedemad4** [comment]
@hydrosquall  sorry for the delay. Just pushed the new code with higher order functions. what do you think?

File: tiingo/wsclient.py
Comment (hydrosquall) [inline tiingo/wsclient.py:91]: Let's make the first argument to `on_error` and `on_close` `self` rather than `ws`, to address the warning provided by LGTM

- https://lgtm.com/projects/g/hydrosquall/tiingo-python/rev/pr-ba63650c8fdd00a84b804ee3327857a1f001cbf6

File: tiingo/wsclient.py
Comment (hydrosquall) [inline tiingo/wsclient.py:44]: We're getting a warning from LGTM about mutating a basic object. I think we can solve it by doing

https://lgtm.com/rules/4840097/

- Make the initial value `None` in the init constructor statement

```python
self.config = {} if config is None else config
```

This should address the warning.

**Review by hydrosquall — CHANGES_REQUESTED** [review]
Almost there! Just 2 more warnings for the LGTM linter to address, and I think we'll be ready to bring this in. The new higher order function approach looks good to me.

File: tiingo/wsclient.py
Comment (Mohamedemad4) [reply tiingo/wsclient.py:91]: that's just going to be confusing. since these functions don't get ```self``` when they are called. and will just receive a ```ws``` instance. i think a better approach would be to just use ```*args``` so as not to confuse future maintainers.BUT i doubt that would stop the LGTM linter from throwing a warning

Comment (hydrosquall) [reply tiingo/wsclient.py:91]: I see your point about not wanting people to think that "self" refers to the Tiingo client instead of the `ws` instance - what do you think about:

- Leaving the function as is, but leaving a comment to explain to future maintainers why the variable is called `ws` on this line. This won't affect LGTM but will help maintainers
- Removing this function from the `TiingoClient` class and defining it as a standalone function above the class - this will solve the LGTM issue.

Comment (Mohamedemad4) [reply tiingo/wsclient.py:91]: let's go with the first solution. i think it's best for this case

Comment (hydrosquall) [reply tiingo/wsclient.py:92]: If you see this in time, can we also add a comment to suppress the warning?

https://lgtm.com/help/lgtm/alert-suppression#python

If not no worries, I'll add this before releasing to PyPi on Friday - this is to avoid future contributors from getting spammed by a warning they can ignore.

File: tiingo/wsclient.py
Comment (hydrosquall) [inline tiingo/wsclient.py:41]: Thanks for updating this line, to complete the change we should also add the following ternary

```
self.config = {} if config is None else config
```

Otherwise, we'll get an error when we try to run `self.config.update` as `update` isn't available on a `None` type.

Comment (hydrosquall) [reply tiingo/wsclient.py:41]: I'm not sure why the travis checks aren't showing in the Github UI, but you can check the test results from the Travis ui here:

https://travis-ci.org/github/hydrosquall/tiingo-python/jobs/733171252

Comment (Mohamedemad4) [reply tiingo/wsclient.py:41]: but like. that should only happen if someone called ```TiingowebsocketClient``` without the config param. python would raise an error. because the param doesn't have a default argument. so it won't even reach the ```self.config.update```.
am i getting this right?

Comment (hydrosquall) [reply tiingo/wsclient.py:41]: Ah, you're right - I have been working mostly in JS recently, and forgot that in Python, if you don't pass in an argument that you'll get an error at runtime (in JS, if you don't pass in an argument, the function runs but assumes that the value of config === `undefined`. I supposed the value of doing the check would have been to guard against someone doing `TiingoWebsocketClient(config=None)`.

Anyways, I think that having a named argument is helpful to callers so they know the meaning of the position, so I guess the full change would look like:

```python
def __init__(config=None, # other args)

    self.config = {} if config is None else config
```

Regarding the failing test, it appears that the error wasn't raised because the test environment contained the TIINGO_API_KEY environment variable. In order to complete the change, you can follow the `delenv` example here:

```
monkeypatch.delenv
```

https://docs.pytest.org/en/4.6.1/monkeypatch.html#monkeypatching-environment-variables

In case you haven't run the unit tests locally before, that can be done by running the following in your terminal when in the root directory, this will be faster than waiting for the tests to run in CI. 

```bash
python setup.py test
```

Comment (hydrosquall) [reply tiingo/wsclient.py:41]: Hi @mohamedemad4, just wanted to check if you might have time to take a look at fixing this test - I think it's the last blocking item that we'd need before we can include your changes in the new release.

Comment (Mohamedemad4) [reply tiingo/wsclient.py]: the link you provided is broken

Comment (Mohamedemad4) [reply tiingo/wsclient.py]: also i don't get why the CI tests are failing? shouldn't the CI be providing it's own TINGO_API_KEY as an env var? or am i missing something?

Comment (hydrosquall) [reply tiingo/wsclient.py]: Hi @Mohamededmad4, thanks for looking into this. The CI environment does have a `TIINGO_API_KEY`. The test is failing because the `TIINGO_API_KEY` is provided, which is why `test_missing_api_key` doesn't work.

https://github.com/hydrosquall/tiingo-python/blob/f6ef473b0d8e951ede092509f17854f7e0ffd053/.github/workflows/python-package.yml#L42-L43

As a result, the code uses that environment variable and we never get to test the "RuntimeError" check for when the config is empty AND nothing is specified in the environment variables. To fix it, you can either change the logic of the client code, or delete the environment variable just inside of that test.

https://github.com/hydrosquall/tiingo-python/blob/be25aa24c3dc2551de164e9ef8bb03a7a30c1f93/tiingo/wsclient.py#L46-L56

Let me know if that's still confusing. The updated link for how to patch environment variables is here:

https://docs.pytest.org/en/stable/monkeypatch.html#monkeypatching-environment-variables

**Comment by Mohamedemad4** [comment]
@hydrosquall I give up. I tried setting up the `lgtm[py/not-named-self]` but they aren't working correctly (or i just used the wrong flags) could you take a look?

**Comment by hydrosquall** [comment]
Hi @Mohamedemad4 , thanks for picking this up again.

> @hydrosquall I give up. I tried setting up the `lgtm[py/not-named-self]` but they aren't working correctly (or i just used the wrong flags) could you take a look?

I think the syntax you used is right, from my reading of the docs

>  alert suppression changes made in a pull request are only applied when that pull request is merged. So, if you suppress an alert in a pull request, LGTM will continue to report the related alert for as long as the pull request is open.

the alert won't go away until the PR is actually merged. 

The more pressing thing to fix is the unit test (see https://github.com/hydrosquall/tiingo-python/runs/3004371828 )

File: tests/test_wsclient.py
Comment (Mohamedemad4) [inline tests/test_wsclient.py:42]: I just realized something. This test might actually hurt other devs writing their own tests on the future if they depend on `TIINGO_API_KEY` being in the `env`. Correct?
@hydrosquall

Comment (Mohamedemad4) [reply tests/test_wsclient.py:42]: Also, This test by deleting the `TIINGO_API_KEY` also fails the test previous to it in the same file. I have no clue how that works.

and I might also add the removing it fixes everything. So I just want to go ahead and delete this test.
is that fine?

**Comment by hydrosquall** [comment]
As mentioned in a previous comment, you don’t want to modify the
environment directly, you probably want to monkeypatch the environment
variable with delenv. This ensures your patch only modifies the state of
this 1 test, and doesn’t have a global effect for other devs or other tests.

https://docs.pytest.org/en/6.2.x/monkeypatch.html

**Comment by Mohamedemad4** [comment]
!ping @hydrosquall. tests are all passing locally now

**Comment by hydrosquall** [comment]
This looks great! Thank you for your efforts here @Mohamedemad4  :)!

**Pull request merged**
